File: atp-tour-angular/src/app/core/services/country.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Country } from 'src/app/models/country.model';
import { environment } from 'src/environments/environment';

/**
 * Service for country data management
 * 
 * @author Lazar
 */
@Injectable({
  providedIn: 'root'
})
export class CountryService {

  private API_URL: string = environment.API_URL;

  /**
   * @constructor
   * 
   * @param {HttpClient} httpClient 
   */
  constructor(private httpClient: HttpClient) { }

  /**
   * Gets countries from the database
   * 
   * @returns {Observable<Country[]>} Array of all countries from the database
    */
  public getCountries(): Observable<Country[]> {
    return this.httpClient.get<Country[]>(this.API_URL + "/country");
  }

  /**
   * Filters countries based on the country name
   * 
   * @param {string} value Filter value
   * @param {Country[]} countries Countries for filtering
   * 
   * @returns {Country[]} Filtered countries
   */
  public filterCountries(value: string | Country, countries: Country[]): Country[] {
    const filterValue = (value instanceof Country) ? value.name : value;
    return countries.filter(option => {
      return option.name.toLowerCase().includes(filterValue)
    })
  }
}

File: atp-tour/src/main/java/com/silab/atptour/service/CountryService.java
package com.silab.atptour.service;

import com.silab.atptour.entity.Country;
import java.util.List;

/**
 * Represents a service containing all the logic for country data management
 *
 * @author Lazar
 */
public interface CountryService {

    /**
     * Retrieves countries from the database
     * 
     * @param name A string representing name of the country
     *
     * @return A {@link List} of countries
     */
    public List<Country> getCountries(String name);
}
